### PR TITLE
Improved handling of 'fresh from console' services

### PR DESCRIPTION
**Pull request by joehan**
### Description
During `firebase init dataconnect`, if there are any services with either no schema or no GQL files n the schema, offer to pull the config from that service.

Also, stop prompting users for connector ID.

### Scenarios Tested
I tested:
 service with no schema:
Screenshot 2024-05-31 at 10 37 56 AM

service with empty schema:
Screenshot 2024-05-31 at 10 13 47 AM

New service (existing flow):
Screenshot 2024-05-31 at 10 38 27 AM

### CODE REVIEW

File: src/init/features/dataconnect/index.ts [hunk 1]
```diff
@@ -18,20 +20,159 @@ const SCHEMA_TEMPLATE = readFileSync(join(TEMPLATE_ROOT, "schema.gql"), "utf8");
...
+    info = await promptForCloudSQLInstance(setup, info);
   }
-  const serviceId = await promptOnce({
-    message: "What ID would you like to use for this service?",
+
+  if (info.cloudSqlDatabase === "") {
+    info = await promptForDatabase(setup, config, info);
+  }
+
+  const defaultConnectionString =
```
Comment (fredzqm) [inline src/init/features/dataconnect/index.ts:52]: Okay, to be a TODO, but eventually we will remove this to a config or prompt when emulator starts.

File: src/init/features/dataconnect/index.ts [hunk 2]
```diff
@@ -18,20 +20,159 @@ const SCHEMA_TEMPLATE = readFileSync(join(TEMPLATE_ROOT, "schema.gql"), "utf8");
...
+      existingServices.map(async (s) => {
+        return {
+          service: s,
+          schema: await getSchema(s.name),
+        };
+      }),
+    );
+    const existingFreshServicesAndSchemas = existingServicesAndSchemas.filter((s) => {
+      return !s.schema?.source.files?.length;
+    });
```
Comment (fredzqm) [inline src/init/features/dataconnect/index.ts:135]: Oh, very interesting, so don't even include not fresh services in the prompt.

Shall we add a TODO to download not fresh services as well?

File: src/init/features/dataconnect/index.ts [hunk 3]
```diff
@@ -18,20 +20,159 @@ const SCHEMA_TEMPLATE = readFileSync(join(TEMPLATE_ROOT, "schema.gql"), "utf8");
...
+      message: "What ID would you like to use for this service?",
+      type: "input",
+      default: "my-service",
+    });
+  }
+  return info;
+}
+
+async function promptForCloudSQLInstance(setup: Setup, info: RequiredInfo): Promise<RequiredInfo> {
+  if (setup.projectId && info.cloudSqlInstanceId) {
```
Comment (fredzqm) [inline src/init/features/dataconnect/index.ts:175]: The caller already checked that `cloudSqlInstanceId` is missing. This branch isn't reachable.

```
  if (info.cloudSqlInstanceId === "") {
    info = await promptForCloudSQLInstance(setup, info);
  }
```

It looks right after Dropping this if-statement.

File: src/init/features/dataconnect/index.ts [hunk 4]
```diff
@@ -69,116 +218,46 @@ export async function doSetup(setup: Setup, config: Config): Promise<void> {
...
 
-    newInstance = true;
-    cloudSqlInstanceId = await promptOnce({
-      message: `What ID would you like to use for your new CloudSQL instance?`,
-      type: "input",
-      default: `fdc-sql`,
-    });
-    locationId = await promptOnce({
+    info.locationId = await promptOnce({
       message: "What location would you use for this instance?",
```
Comment (fredzqm) [inline src/init/features/dataconnect/index.ts:222]: nit: This is both the location for FDC service and CSQL instance.

It's only needed for new service and new instance.
Maybe say "what location would you use?"

File: src/init/features/dataconnect/index.ts [hunk 5]
```diff
@@ -18,20 +20,159 @@ const SCHEMA_TEMPLATE = readFileSync(join(TEMPLATE_ROOT, "schema.gql"), "utf8");
...
+      default: "my-service",
+    });
+  }
+  return info;
+}
+
+async function promptForCloudSQLInstance(setup: Setup, info: RequiredInfo): Promise<RequiredInfo> {
+  if (setup.projectId && info.cloudSqlInstanceId) {
     const instances = await cloudsql.listInstances(setup.projectId);
     const choices = instances.map((i) => {
```
Comment (fredzqm) [inline src/init/features/dataconnect/index.ts:177]: If the service already exists, we know its region, so should filter the Cloud SQL list here by that region.

**Review by fredzqm — CHANGES_REQUESTED** [review]

File: src/init/features/dataconnect/index.ts (hunk 2 above)
Comment (joehan) [reply src/init/features/dataconnect/index.ts:135]: Yep - that was the intention of the TODO on L123

**Review by yuchenshi — APPROVED** [review]
Approval for the flow of interactions. Please address Fred's comments though

File: src/init/features/dataconnect/index.ts (hunk 5 above)
Comment (joehan) [reply src/init/features/dataconnect/index.ts:177]: Done.

File: src/init/features/dataconnect/index.ts (hunk 3 above)
Comment (joehan) [reply src/init/features/dataconnect/index.ts:175]: Done.

File: src/init/features/dataconnect/index.ts (hunk 4 above)
Comment (joehan) [reply src/init/features/dataconnect/index.ts:222]: Done.

File: src/init/features/dataconnect/index.ts [hunk 6]
```diff
@@ -182,3 +117,154 @@ function subValues(
...
+      return { name: i.name, value: i.name, location: i.region };
+    });
+
+    const freeTrialInstanceId = await checkForFreeTrialInstance(setup.projectId);
+    if (!freeTrialInstanceId) {
+      choices.push({ name: "Create a new instance", value: "", location: "" });
+    }
+    // If we've already chosen a region (ie service already exists), only list instances from that region.
+    choices = choices.filter((c) => info.locationId === "" || info.locationId === c.location);
+    if (instances.length) {
```
Comment (fredzqm) [inline src/init/features/dataconnect/index.ts:188]: Very minor bug here.

```
    choices = choices.filter((c) => info.locationId === "" || info.locationId === c.location);
    if (choices.length) {
      if (!freeTrialInstanceId) {
        choices.push({ name: "Create a new instance", value: "", location: "" });
      }
      // then prompt
    }
```

If the filter list is empty, creating new instance is the only option, no need to prompt

**Review by fredzqm — APPROVED** [review]

File: src/init/features/dataconnect/index.ts (hunk 6 above)
Comment (fredzqm) [reply src/init/features/dataconnect/index.ts:188]: 😓 oh, did I just trigger auto-submit?

**Comment by fredzqm** [comment]
Emm, how does auto-merge work?

Is there a way to approve a PR and leave a final nit without triggering it? Critique won't trigger auto-submit when there are unresolved comments.

**Comment by joehan** [comment]
All good - will clean up the empty case in a follow up PR. TLDR is the autosubmit submits as soon as all the requirements pass. On firebase-tools, those requirements are 1 approving review, no requested changes reviews, and all passing required tests.

**Pull request merged**